File: examples/clj/src/com/rpl/agent/basic/rama_module_agent.clj
(ns com.rpl.agent.basic.rama-module-agent
  "Demonstrates using Rama module directly instead of defagentmodule.

  Features demonstrated:
  - rama/module: Define a Rama module directly
  - agents-topology: Create agent topology manually
  - underlying-stream-topology: Access the underlying StreamTopology
  - declare-depot: Declare a Rama depot alongside agents
  - <<sources: Process depot streams with Rama's stream processing
  - new-agent: Create an agent within the topology
  - node: Define agent node
  - define-agents!: Explicitly finalize agent definitions
  
  This example shows how to integrate agent-o-rama with full Rama features
  when you need access to depots, stream processing, or other Rama primitives."
  (:require
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]))

;;; Rama module with agents and depot
(def RamaModuleAgent
  (rama/module
   {:module-name "RamaModuleAgent"}
   [setup topologies]
   ;; Declare a depot to demonstrate Rama feature integration
   (rama/declare-depot setup *example-depot (rama/hash-by identity))

   ;; Create agents topology
   (let [topology (aor/agents-topology setup topologies)
         ;; Access underlying stream topology for Rama features
         s-topology (aor/underlying-stream-topology topology)]

     ;; Use Rama's stream processing alongside agents
     (rama/<<sources
      s-topology
      (rama/source> *example-depot :> *value)
      (println "Depot received:" *value))

     ;; Define a simple feedback agent
     (-> topology
         (aor/new-agent "FeedbackAgent")
         (aor/node
          "process-feedback"
          nil
          (fn [agent-node feedback-text]
            ;; Process feedback and return success response
            (let [response {:status :success
                            :message (str "Processed: " feedback-text)
                            :length (count feedback-text)}]
              (aor/result! agent-node response)))))

     ;; Finalize agent definitions
     (aor/define-agents! topology))))

(defn -main
  "Run the rama module agent example"
  [& _args]
  ;; Create in-process cluster
  (with-open [ipc (rtest/create-ipc)]
    ;; Launch the Rama module
    (rtest/launch-module! ipc RamaModuleAgent {:tasks 1 :threads 1})

    (let [module-name (rama/get-module-name RamaModuleAgent)]
      ;; Get the depot for demonstration
      (let [depot (rama/foreign-depot ipc module-name "*example-depot")]
        ;; Append to depot to show it's working
        (println "\n=== Depot Integration Demo ===")
        (rama/foreign-append! depot "Example data 1")
        (rama/foreign-append! depot "Example data 2")
        (Thread/sleep 100)) ; Allow time for depot processing

      ;; Get agent manager and client
      (let [manager (aor/agent-manager ipc module-name)
            agent (aor/agent-client manager "FeedbackAgent")]

        (println "\n=== Agent Results ===")
        (println "Available agents:" (aor/agent-names manager))

        ;; Invoke agent with sample feedback
        (let [result1 (aor/agent-invoke agent "Great product!")
              result2 (aor/agent-invoke agent "Needs improvement")]
          (println "Feedback 1:" result1)
          (println "Feedback 2:" result2))))))

File: examples/java/src/main/java/com/rpl/agent/basic/RamaModuleAgent.java
package com.rpl.agent.basic;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.agentorama.AgentNode;
import com.rpl.agentorama.AgentTopology;
import com.rpl.rama.Depot;
import com.rpl.rama.RamaModule;
import com.rpl.rama.module.StreamTopology;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.HashMap;

/**
 * Java example demonstrating direct Rama module usage instead of AgentModule.
 *
 * <p>This example demonstrates:
 *
 * <ul>
 *   <li>Implementing RamaModule directly (not extending AgentModule)
 *   <li>Creating AgentTopology manually via AgentTopology.create()
 *   <li>Accessing StreamTopology via getStreamTopology()
 *   <li>Declaring Rama depots alongside agents
 *   <li>Using Rama's stream processing with agents
 *   <li>Explicitly calling topology.define()
 * </ul>
 *
 * <p>This approach allows integration of agent-o-rama with full Rama features when you need access
 * to depots, stream processing, or other Rama primitives.
 */
public class RamaModuleAgent {

  /**
   * Direct Rama Module implementation showing integration with full Rama features.
   *
   * <p>This module demonstrates manual topology creation and depot integration, providing access to
   * the complete Rama feature set alongside agent-o-rama agents.
   */
  public static class RamaModule implements com.rpl.rama.RamaModule {

    public RamaModule() {
      super();
    }

    @Override
    public String getModuleName() {
      return "RamaModuleAgent";
    }

    @Override
    public void define(Setup setup, Topologies topologies) {
      // Declare a depot to demonstrate Rama feature integration
      setup.declareDepot("*example-depot", Depot.random());

      // Create agents topology manually
      AgentTopology topology = AgentTopology.create(setup, topologies);

      // Access underlying stream topology for Rama features (available for custom processing)
      StreamTopology streamTopology = topology.getStreamTopology();

      // Note: Stream topology can be used here for custom Rama stream processing
      // For simplicity, this example focuses on the agent topology integration

      // Define a simple feedback agent
      topology
          .newAgent("FeedbackAgent")
          .node("process-feedback", null, (AgentNode agentNode, String feedbackText) -> {
            // Process feedback and return success response
            HashMap<String, Object> response = new HashMap<>();
            response.put("status", "success");
            response.put("message", "Processed: " + feedbackText);
            response.put("length", feedbackText.length());

            agentNode.result(response);
          });

      // Explicitly finalize agent definitions
      topology.define();
    }
  }


  public static void main(String[] args) throws Exception {
    System.out.println("Starting Rama Module Agent Example...");

    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Launch the Rama module
      RamaModule module = new RamaModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);

      System.out.println("\n=== Agent Results ===");
      System.out.println("Available agents: " + manager.getAgentNames());

      // Note: The depot "*example-depot" is declared and available for use
      // For demonstration purposes, we focus on the agent functionality

      // Get client for our agent
      AgentClient agent = manager.getAgentClient("FeedbackAgent");

      // Invoke agent with sample feedback
      Object result1 = agent.invoke("Great product!");
      Object result2 = agent.invoke("Needs improvement");

      System.out.println("Feedback 1: " + result1);
      System.out.println("Feedback 2: " + result2);
    }
  }
}
